src/raft/replication.c

```diff
@@ -880,15 +887,60 @@ static void sendAppendEntriesResult(
 	}
 }
 
+/**
+ * State machine for handling of AppendEntries on the follower side.
+ */
+enum {
+	AF_START,
+	AF_DONE,
+	AF_FAILED,
+	AF_NR,
+};
+
+static struct sm_conf af_states[AF_NR] = {
+	[AF_START] = {
+		.name = "start",
+		.allowed = BITS(AF_DONE)|BITS(AF_FAILED),
+		.flags = SM_INITIAL,
+	},
+	[AF_DONE] = {
+		.name = "done",
+		.flags = SM_FINAL,
+	},
+	[AF_FAILED] = {
+		.name = "failed",
+		.flags = SM_FAILURE|SM_FINAL,
+	},
+};
+
+static bool af_invariant(const struct sm *sm, int prev)
+{
+	(void)sm;
+	(void)prev;
+	return true;
+}
+
 /* Context for a write log entries request that was submitted by a follower. */
 struct appendFollower
 {
 	struct raft *raft; /* Instance that has submitted the request */
 	raft_index index;  /* Index of the first entry in the request. */
 	struct raft_append_entries args;
 	struct raft_io_append req;
+	struct sm sm;
 };
 
+static void append_follower_done(struct appendFollower *req, int status)
+{
+	if (status == 0) {
+		sm_move(&req->sm, AF_DONE);
+	} else {
+		sm_fail(&req->sm, AF_FAILED, status);
+	}
```

Comment:
Such a construction can be seen elsewhere in the patch. I'd expect to look at it this following way:
```
sm_move_rc(&req->sm, status == 0 ? DONE : FAILED, status);
```
So maybe it's worth defining a new function in sm.c/h

Reply:
Good idea.